circuit/src/gadgets/range.rs

```diff
@@ -0,0 +1,460 @@
+			let bitshift_selector = meta.query_selector(bitshift_selector);
+			let shifted_word = meta.query_advice(shifted_word_column, Rotation::cur());
+
+			vec![(bitshift_selector * shifted_word, mock_common.table)]
+		});
+
+		// For short lookups, check that the word has been shifted by the correct number
+		// of bits. https://p.z.cash/halo2-0.1:decompose-short-lookup
+		meta.create_gate("Short word S bits limit", |meta| {
+			let bitshift_selector = meta.query_selector(bitshift_selector);
+			let word = meta.query_advice(word_column, Rotation::cur());
+			let shifted_word = meta.query_advice(shifted_word_column, Rotation::cur());
+
+			let two_pow_k_min_s = F::from(1 << (K - S));
+
+			// shifted_word = word * 2^{K-S}
+			Constraints::with_selector(
+				bitshift_selector,
+				Some(word * two_pow_k_min_s - shifted_word),
+			)
+		});
+
+		bitshift_selector
+	}
+
+	fn synthesize(
+		self, mock_common: &MockCommonConfig, selector: &Selector, mut layouter: impl Layouter<F>,
+	) -> Result<Self::Output, Error> {
+		layouter.assign_region(
+			|| "short word check chip",
+			|region| {
+				let mut ctx = RegionCtx::new(region, 0);
+				ctx.enable(selector.clone())?;
+
+				// Assign original value
+				let assigned_x = ctx.copy_assign(mock_common.common.advice[0], self.x.clone())?;
+
+				// Assign shifted value
+				let shifted_word = self.x.value().cloned() * Value::known(F::from(1 << (K - S)));
+				ctx.assign_advice(mock_common.common.advice[1], shifted_word)?;
+
+				Ok(assigned_x)
+			},
+		)
+	}
+}
+
+/// Range check chip using lookup table.
+///
+/// Check if the target value is `K * N` bits long.
+#[derive(Debug, Clone)]
+pub struct LookupRangeCheckChip<F: FieldExt, const K: usize, const N: usize> {
+	x: AssignedCell<F, F>,
+}
+
+impl<F: FieldExt, const K: usize, const N: usize> LookupRangeCheckChip<F, K, N> {
+	/// Construct a new instance
+	pub fn new(x: AssignedCell<F, F>) -> Self {
+		Self { x }
+	}
+}
+
+impl<F: FieldExt, const K: usize, const N: usize> MockChip<F> for LookupRangeCheckChip<F, K, N> {
+	type Output = (AssignedCell<F, F>, AssignedCell<F, F>);
+
+	fn configure(mock_common: &MockCommonConfig, meta: &mut ConstraintSystem<F>) -> Selector {
+		let running_sum_selector = meta.complex_selector();
+
+		let running_sum = mock_common.common.advice[0];
+
+		meta.lookup("running_sum check", |meta| {
+			let running_sum_selector = meta.query_selector(running_sum_selector);
+			let z_cur = meta.query_advice(running_sum, Rotation::cur());
+
+			/*
+				Example of 16 bits number:
+				2 = 16 / 8
+				0x01ff
+
+				z_0 1111111100000001 (little-endian)
+
+				z_1 00000001  		11111111 = 1111111100000001 - 00000001 * (000000001)
+									a_0      = z_0 				- z_1 	   * (1 << 8)
+				z_2 00000000   		00000001 = 00000001 - 0   * (000000001)
+									a_1 	 = z_1 		- z_2 * (1 << 8)
+
+				In the case of a running sum decomposition, we recover the word from
+				the difference of the running sums:
+					z_i = 2^{K}⋅z_{i + 1} + a_i
```

Comment:
Can you elaborate on how you got this equation.
Because [here](https://github.com/eigen-trust/protocol/pull/169/files#diff-11513d52971f0935b0b5cf0c36e0b15b6d9ef1595f05ed0915140a3cd5175dfbR175), the equation is:
`z_i = (z_{i - 1} - a_{i - 1}) / 2^K`